Maven artifact cannot be downloaded while bootstrapping app

**Describe the bug**
When running an `@QaurkusTest` annotated test the Maven resolver configured by Quarkus fails to download an artifact from our Maven repository. The remote Maven repository is hosted by GitLab and requires a security token to be provided as an HTTP header. This has been configured in Maven's `settings.xml`, but it isn't picked up by Quarkus.

**Expected behavior**
Quarkus should be able to download the Maven artifact just like the Maven build does it.

**Actual behavior**
When running the test using Maven the test fails if the deployment artifact of a custom Quarkus extension isn't already present in the local Maven repository, because it can't download it from the remote repository. When running the test the following error gets logged:
```
java.lang.RuntimeException: io.quarkus.bootstrap.BootstrapException: Failed to create the application model for ch.mobi.mcs:mcs-b2esimple-service::jar:1.0.0-SNAPSHOT
Caused by: io.quarkus.bootstrap.BootstrapException: Failed to create the application model for ch.mobi.mcs:mcs-b2esimple-service::jar:1.0.0-SNAPSHOT
Caused by: io.quarkus.bootstrap.resolver.AppModelResolverException: Failed to inject extension deployment dependencies for ch.mobi.mcs:mcs-b2esimple-service:jar:1.0.0-SNAPSHOT
Caused by: io.quarkus.bootstrap.BootstrapDependencyProcessingException: No dependencies collected for Quarkus extension deployment artifact ch.mobi.gluon:gluon-commands-deployment:jar:0.1.424 while at least the corresponding runtime artifact ch.mobi.gluon:gluon-commands:jar:0.1.424 is expected
```

I suspect the problem is in `BootstrapMavenContext`. In `newRepositorySystemSession()` the code adds `XmlPlexusConfiguration` config properties for the Maven repository servers as configured in `settings.xml`, but the corresponding `WagonConfigurator` service implementation `PlexusWagonConfigurator` is not registered in `newRepositorySystem()`. Registering that service implementation might fix the issue, but I have not had time to verify that yet.

**To Reproduce**
Not so easy... I can fill this out at a later stage, if still required.

**Configuration**
In `~/.m2/settings.xml` the additional HTTP header is configured like this:
```xml
    <servers>
        <server>
            <id>gitlab-maven</id>
            <configuration>
                <httpHeaders>
                    <property>
                        <name>Private-Token</name>
                        <value>XXXXXXXXXXXXXXXXXX</value>
                    </property>
                </httpHeaders>
            </configuration>
        </server>
    </servers>
```

**Screenshots**

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: MINGW64_NT-10.0-19624
 - Output of `java -version`: openjdk version "11.0.6" 2020-01-14 LTS
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 1.4.2.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Maven 3.6.3

**Additional context**

--- a/independent-projects/bootstrap/core/pom.xml
+++ b/independent-projects/bootstrap/core/pom.xml
@@ -23,6 +23,10 @@
             <artifactId>maven-embedder</artifactId>
         </dependency>
         <dependency>
+            <groupId>org.eclipse.sisu</groupId>
+            <artifactId>org.eclipse.sisu.plexus</artifactId>
+        </dependency>
+        <dependency>
             <groupId>org.apache.maven</groupId>
             <artifactId>maven-settings-builder</artifactId>
         </dependency>

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/BootstrapMavenContext.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/BootstrapMavenContext.java
@@ -70,6 +70,7 @@ import org.eclipse.aether.resolution.ArtifactDescriptorException;
 import org.eclipse.aether.resolution.ArtifactDescriptorRequest;
 import org.eclipse.aether.spi.connector.RepositoryConnectorFactory;
 import org.eclipse.aether.spi.connector.transport.TransporterFactory;
+import org.eclipse.aether.transport.wagon.WagonConfigurator;
 import org.eclipse.aether.transport.wagon.WagonProvider;
 import org.eclipse.aether.transport.wagon.WagonTransporterFactory;
 import org.eclipse.aether.util.repository.AuthenticationBuilder;
@@ -560,6 +561,7 @@ public class BootstrapMavenContext {
         if (!isOffline()) {
             locator.addService(RepositoryConnectorFactory.class, BasicRepositoryConnectorFactory.class);
             locator.addService(TransporterFactory.class, WagonTransporterFactory.class);
+            locator.setServices(WagonConfigurator.class, new BootstrapWagonConfigurator());
             locator.setServices(WagonProvider.class, new BootstrapWagonProvider());
         }
         locator.setServices(ModelBuilder.class, new MavenModelBuilder(workspace, getCliOptions(),

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/BootstrapWagonConfigurator.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/resolver/maven/BootstrapWagonConfigurator.java
@@ -1,0 +1,36 @@
+package io.quarkus.bootstrap.resolver.maven;
+
+import org.apache.maven.wagon.Wagon;
+import org.codehaus.plexus.component.configurator.converters.composite.ObjectWithFieldsConverter;
+import org.codehaus.plexus.component.configurator.converters.lookup.ConverterLookup;
+import org.codehaus.plexus.component.configurator.converters.lookup.DefaultConverterLookup;
+import org.codehaus.plexus.component.configurator.expression.DefaultExpressionEvaluator;
+import org.codehaus.plexus.component.configurator.expression.ExpressionEvaluator;
+import org.codehaus.plexus.configuration.PlexusConfiguration;
+import org.codehaus.plexus.configuration.xml.XmlPlexusConfiguration;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+import org.eclipse.aether.transport.wagon.WagonConfigurator;
+
+public class BootstrapWagonConfigurator implements WagonConfigurator {
+
+    private static final ExpressionEvaluator DEFAULT_EXPRESSION_EVALUATOR = new DefaultExpressionEvaluator();
+
+    protected ConverterLookup converterLookup = new DefaultConverterLookup();
+
+    @Override
+    public void configure(Wagon wagon, Object configuration) throws Exception {
+        PlexusConfiguration config;
+        if (configuration instanceof PlexusConfiguration) {
+            config = (PlexusConfiguration) configuration;
+        } else if (configuration instanceof Xpp3Dom) {
+            config = new XmlPlexusConfiguration((Xpp3Dom) configuration);
+        } else if (configuration == null) {
+            return;
+        } else {
+            throw new IllegalArgumentException("unexpected configuration type: "
+                    + configuration.getClass().getName());
+        }
+        final ObjectWithFieldsConverter converter = new ObjectWithFieldsConverter();
+        converter.processConfiguration(converterLookup, wagon, null, config, DEFAULT_EXPRESSION_EVALUATOR, null);
+    }
+}

--- a/independent-projects/bootstrap/pom.xml
+++ b/independent-projects/bootstrap/pom.xml
@@ -45,6 +45,7 @@
         <shrinkwrap-depchain.version>1.2.6</shrinkwrap-depchain.version>
         <jboss-logmanager-embedded.version>1.0.4</jboss-logmanager-embedded.version>
         <graal-sdk.version>19.3.1</graal-sdk.version>
+        <plexus-classworlds.version>2.6.0</plexus-classworlds.version> <!-- not actually used but ClassRealm class is referenced from the API used in BootstrapWagonConfigurator -->
     </properties>
     <modules>
         <module>core</module>
@@ -144,10 +145,6 @@
                     <exclusion>
                         <groupId>javax.enterprise</groupId>
                         <artifactId>cdi-api</artifactId>
-                    </exclusion>
-                    <exclusion>
-                        <groupId>org.codehaus.plexus</groupId>
-                        <artifactId>plexus-classworlds</artifactId>
                     </exclusion>
                     <exclusion>
                         <groupId>org.checkerframework</groupId>
@@ -270,16 +267,8 @@
                 <version>${sisu.version}</version>
                 <exclusions>
                     <exclusion>
-                        <groupId>org.codehaus.plexus</groupId>
-                        <artifactId>plexus-component-annotations</artifactId>
-                    </exclusion>
-                    <exclusion>
-                        <groupId>org.codehaus.plexus</groupId>
-                        <artifactId>plexus-utils</artifactId>
-                    </exclusion>
-                    <exclusion>
-                        <groupId>org.codehaus.plexus</groupId>
-                        <artifactId>plexus-classworlds</artifactId>
+                        <groupId>*</groupId>
+                        <artifactId>*</artifactId>
                     </exclusion>
                 </exclusions>
             </dependency>
@@ -302,6 +291,17 @@
                 <groupId>org.jboss.logging</groupId>
                 <artifactId>commons-logging-jboss-logging</artifactId>
                 <version>${commons-logging-jboss-logging.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.codehaus.plexus</groupId>
+                <artifactId>plexus-classworlds</artifactId>
+                <version>${plexus-classworlds.version}</version>
+                <exclusions>
+                    <exclusion>
+                        <groupId>*</groupId>
+                        <artifactId>*</artifactId>
+                    </exclusion>
+                </exclusions>
             </dependency>
             <dependency>
                 <groupId>org.junit.jupiter</groupId>

